### ISSUE TITLE
void-function flycheck-define-checker when using use-package

### ISSUE THREAD

**Issue by suntong** [flycheck/flycheck#1889]
This template is for **user support**.  If you are reporting an actual Flycheck bug, please use the relevant issue template instead.

## Checklist
- [ ] I have searched existing issues and StackOverflow for solutions.
- [ ] I have read the [Flycheck manual][].
- [ ] I have read the [Troubleshooting Guide][].
- [ ] I have checked my issue has not been reported yet.

## Problem description

Following up on #567, I'm not loading `flycheck` with `(require 'flycheck)` but `use-package` instead. I am getting the exact same error as in #567, i.e., with `--debug-init`:

```
Debugger entered--Lisp error: (void-function flycheck-define-checker)
  (flycheck-define-checker clang++-ledger "Clang++ checker for Ledger" :command '("clang++" "-Wall" "-fsyntax-only" "-I/Users/johnw/Products/ledger/debug" "-I../lib" "-I../lib/utfcpp/source" "-I/System/Library/Frameworks/Python.framework/Vers..." "-include" "system.hh" "-c" source-inplace) :error-patterns '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s..." warning) ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*..." error)) :modes 'c++-mode :predicate '(string-match "/ledger/" (buffer-file-name)))
  (progn (flycheck-define-checker clang++-ledger "Clang++ checker for Ledger" :command '("clang++" "-Wall" "-fsyntax-only" "-I/Users/johnw/Products/ledger/debug" "-I../lib" "-I../lib/utfcpp/source" "-I/System/Library/Frameworks/Python.framework/Vers..." "-include" "system.hh" "-c" source-inplace) :error-patterns '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s..." warning) ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*..." error)) :modes 'c++-mode :predicate '(string-match "/ledger/" (buffer-file-name))) (setq flycheck-checkers (cons 'clang++-ledger flycheck-checkers)))
  (condition-case err (progn (flycheck-define-checker clang++-ledger "Clang++ checker for Ledger" :command '("clang++" "-Wall" "-fsyntax-only" "-I/Users/johnw/Products/ledger/debug" "-I../lib" "-I../lib/utfcpp/source" "-I/System/Library/Frameworks/Python.framework/Vers..." "-include" "system.hh" "-c" source-inplace) :error-patterns '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s..." warning) ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*..." error)) :modes 'c++-mode :predicate '(string-match "/ledger/" (buffer-file-name))) (setq flycheck-checkers (cons 'clang++-ledger flycheck-checkers))) ((debug error) (funcall use-package--warning2 :init err)))
  (progn (condition-case err (progn (flycheck-define-checker clang++-ledger "Clang++ checker for Ledger" :command '("clang++" "-Wall" "-fsyntax-only" "-I/Users/johnw/Products/ledger/debug" "-I../lib" "-I../lib/utfcpp/source" "-I/System/Library/Frameworks/Python.framework/Vers..." "-include" "system.hh" "-c" source-inplace) :error-patterns '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s..." warning) ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*..." error)) :modes 'c++-mode :predicate '(string-match "/ledger/" (buffer-file-name))) (setq flycheck-checkers (cons 'clang++-ledger flycheck-checkers))) ((debug error) (funcall use-package--warning2 :init err))) (if (not (require 'flycheck nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck) :error) (condition-case err (progn (defalias 'flycheck-show-error-at-point-soon 'flycheck-show-error-at-point) t) ((debug error) (funcall use-package--warning2 :config err)))))
  (condition-case err (progn (condition-case err (progn (flycheck-define-checker clang++-ledger "Clang++ checker for Ledger" :command '("clang++" "-Wall" "-fsyntax-only" "-I/Users/johnw/Products/ledger/debug" "-I../lib" "-I../lib/utfcpp/source" "-I/System/Library/Frameworks/Python.framework/Vers..." "-include" "system.hh" "-c" source-inplace) :error-patterns '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s..." warning) ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*..." error)) :modes 'c++-mode :predicate '(string-match "/ledger/" (buffer-file-name))) (setq flycheck-checkers (cons 'clang++-ledger flycheck-checkers))) ((debug error) (funcall use-package--warning2 :init err))) (if (not (require 'flycheck nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck) :error) (condition-case err (progn (defalias 'flycheck-show-error-at-point-soon 'flycheck-show-error-at-point) t) ((debug error) (funcall use-package--warning2 :config err))))) ((debug error) (funcall use-package--warning2 :catch err)))
  eval-buffer(#<buffer  *load*-123847> nil ...
```

## Steps to reproduce

put the following in init:

```el
;;;_* flycheck
(use-package flycheck
  :config
  (defalias 'flycheck-show-error-at-point-soon 'flycheck-show-error-at-point)
  :init
  (progn
    (flycheck-define-checker clang++-ledger
      "Clang++ checker for Ledger"
      :command
      '("clang++" "-Wall" "-fsyntax-only"
        "-I/Users/johnw/Products/ledger/debug" "-I../lib"
        "-I../lib/utfcpp/source"
        "-I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7"
        "-include" "system.hh" "-c" source-inplace)
      :error-patterns
      '(("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): warning:\\s-*\\(?4:.*\\)"
         warning)
        ("^\\(?1:.*\\):\\(?2:[0-9]+\\):\\(?3:[0-9]+\\): error:\\s-*\\(?4:.*\\)"
         error))
      :modes 'c++-mode
      :predicate '(string-match "/ledger/" (buffer-file-name)))

    (push 'clang++-ledger flycheck-checkers)))
```

Start/restart emacs with `--debug-init`.

- [ ] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior

It works.

## System configuration

`M-x flycheck-verify-setup` says "no match"

Emacs configuration:
- [ ] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes

Can someone give me a working `flycheck` setup using `use-package` please as I don't know elisp except copying from working code? thx.

[Flycheck manual]:  https://www.flycheck.org/en/latest/user/troubleshooting.html
[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by suntong** [comment]
Decided to remove all `clang` related setting, as:

```el
(use-package flycheck
  :ensure t
  :config
  (defalias 'flycheck-show-error-at-point-soon 'flycheck-show-error-at-point)
  :init (global-flycheck-mode))
```

but still appreciate a working example to define `flycheck-define-checker` within `use-package`.
